kafka-connect-s3/src/main/java/io/confluent/connect/s3/S3SinkTaskV2.java

```diff
@@ -0,0 +1,86 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import com.hotstar.kafka.connect.transforms.ProtoToPayloadTransform;
+
+public class S3SinkTaskV2 extends S3SinkTask {
+    private ExecutorService executorService;
+    public static final long THREAD_KEEP_ALIVE_TIME_SECONDS = 60L;
+    private static final Logger LOGGER = LoggerFactory.getLogger(S3SinkTaskV2.class);
+
+    private static final ProtoToPayloadTransform.Value<SinkRecord> protoToPayloadTransform = new ProtoToPayloadTransform.Value<>();
+    private static final CoerceToSegmentPayload<SinkRecord> coerceToSegmentPayloadTransform = new CoerceToSegmentPayload.Value<>();
+
+    static {
+        // This is needed to initialise the statsD client with default values.
+        protoToPayloadTransform.configure(new HashMap<>());
+
+    }
+
+
+    @Override
+    public void start(Map<String, String> props) {
+        super.start(props);
+        int transformThreadPoolMaxSize = Integer.parseInt(props.get("transform.threads.max"));
+        this.executorService = new ThreadPoolExecutor(1, transformThreadPoolMaxSize,
+                THREAD_KEEP_ALIVE_TIME_SECONDS,
+                TimeUnit.SECONDS,
+                new SynchronousQueue<>(),
+                new ThreadPoolExecutor.CallerRunsPolicy()
+        );
+    }
+
+    @Override
+    public void put(Collection<SinkRecord> records) throws ConnectException {
+        transformInParallel(records);
+        super.put(records);
+    }
+
+    private Collection<SinkRecord> transformInParallel(Collection<SinkRecord> sinkRecords) {
+        int batchSize = sinkRecords.size();
+        SinkRecord[] kafkaRecordArray = sinkRecords.toArray(new SinkRecord[0]);
+        List<Future<SinkRecord>> futureList = new ArrayList<>(batchSize);
+//        LOGGER.debug("In transformInParallel Method.");
+        for (int i = 0; i < batchSize; i++) {
+            int recordNumber = i;
+            futureList.add(executorService.submit((Callable) () -> transformProtoToEventPayload(kafkaRecordArray[recordNumber])));
+        }
+        List<SinkRecord> transformedRecords = new ArrayList<>(batchSize);
+        futureList.stream().forEach(future -> {
+            try {
+                SinkRecord transformedRecord = future.get();
+                transformedRecords.add(transformedRecord);
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            } catch (ExecutionException e) {
+                throw new RuntimeException(e);
+            }
+        });
```

Comment:
why not `parallelStream()`?

Reply:
Is there a way to insert in a list parallely without taking a hit on performance ?

There is a 
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/CopyOnWriteArrayList.html
which copies the entire list on every insert. So this will be inefficient.

Collections.synchronizedList() will give me a synchronized list, but, it will also implement some sort of locking which will imply that one insert operation would wait for the other to complete.

So, not seeing much of a benefit by using parallel stream here.